docs/rfc/rfc-001-end-to-end-testing.md

```diff
@@ -0,0 +1,134 @@
+
+The test suite will do black-box testing against a single node's RPC service. We will be testing the behavior of the network as a whole, e.g. that a fast synced node correctly catches up to the chain head and serves basic block data via RPC. Thus the tests will not send e.g. P2P messages or examine the node database, as these are considered internal implementation details - if the network behaves correctly, presumably the internal components function correctly. Comprehensive component testing (e.g. each and every RPC method parameter) should be done via unit/integration tests.
+
+The tests must take into account the node configuration (e.g. some nodes may be pruned, others may not be validators), and should somehow be provided access to expected data (i.e. complete block headers for the entire chain).
+
+The test suite should use the Tendermint RPC client and the Tendermint light client, to exercise the client code as well.
+
+### Implementation Considerations
+
+The testnets should run in Docker Compose, both locally and in CI. This makes it easier to reproduce test failures locally. Supporting multiple test-runners (e.g. on VMs or Kubernetes) is out of scope. The same image should be used for all tests, with configuration passed via a mounted volume.
+
+There does not appear to be any off-the-shelf solutions that would do this for us, so we will have to roll our own on top of Docker Compose. This gives us more flexibility, but is estimated to be a few weeks of work.
+
+Testnets should be configured via a YAML file. These are used as inputs for the test runner, which e.g. generates Docker Compose configurations from them. An additional layer on top should generate these testnet configurations from a YAML file that specifies all the option combinations to test.
+
+Comprehensive testnets should run against master nightly. However, a small subset of representative testnets should run for each pull request, e.g. a four-node IPv4 network with state sync and fast sync.
+
+Tests should be written using the standard Go test framework (and e.g. Testify), with a helper function to fetch info from the test configuration. The test runner will run the tests separately for each network node, and the test must vary its expectations based on the node's configuration.
+
+It should be possible to launch a specific testnet and run individual test cases from the IDE or local terminal against a it.
+
+If possible, the existing `testnet` command should be extended to set up the network topologies needed by the end-to-end tests.
+
+## Status
+
+Proposed
+
+## Consequences
+
+### Positive
+
+- Comprehensive end-to-end test coverage of basic Tendermint functionality
+
+- Test environments can easily be reproduced locally and debugged via standard tooling
+
+### Negative
+
+- Limited coverage of consensus correctness testing (e.g. Jepsen)
+
+- No coverage of malicious or adversarial behavior
+
+- Have to roll our own test framework, which takes engineering resources
+
```

Comment:
Other negatives include cost and operations overhead, and (potentially) slower merge times, depending on what tests get run on each PR.